Allow audio fallback to be "None"

jstebbins opened the pull request:
When audio fallback is "None", a failure to do passthru will result in
no output audio track being added.

If your GUI uses hb_preset_job_init() and/or hb_preset_job_add_audio() to populate your job with audio, then simply setting preset key "AudioEncoderFallback" to "none" is all that is required from the GUI.  If you are populating the job yourself, then adding a "none" option is pretty much all on you.

jstebbins committed:
Allow audio fallback to be "None"

When audio fallback is "None", a failure to do passthru will result in
no output audio track being added.

galad87 commented:
Can you make hb_audio_encoder_get_name() work with "None" too? MacGui uses it to translate between the int codec and the codec name.

jstebbins commented:
@galad87 working on it.  I'm looking over the ramifications of enabling HB_ACODEC_NONE in HB_ACODEC_MASK which would make all those accessor functions work without adding special casing.

Question, does it make sense to make "None" implied when fallback isn't specified?  This would be a CLI behavioural change.

bradleysepos commented:
The official presets, even CLI default, have fallback specified. So I'm not sure when this condition would be met.

jstebbins commented:
@bradleysepos good point.  I think I'll simplify the underlying code so that it doesn't automatically fall back to aac then since this should only happen in an error condition where the preset value is invalid.

An implementation note that you all might need to know.  I've set the "allowed muxers" to 0 for HB_ACODEC_NONE.  The intent is that this will automatically filter the "None" option out of your regular audio encoder dropdowns.  Let me know if this isn't the case and if there is something else I could do to simplify integrating this into the UIs.

bradleysepos commented:
> I think I'll simplify the underlying code so that it doesn't automatically fall back to aac then since this should only happen in an error condition where the preset value is invalid.

👍

jstebbins committed:
make HB_ACODEC_NONE visible through hb_audio_encoder_get_*

jstebbins committed:
simplify audio autopassthru fallback logic

Drop track when fallback codec is invalid instead of falling back to a
default.  Since all presets have a fallback set, the default fallback
condition would only ever be triggered by an invalid setting.

jstebbins committed:
LinGui: only show "None" in fallback code list

galad87 commented:
Should it remove a track if for example the source is AAC, DTS-HD passthru is selected and the fallback is AAC? 1.0.x keeps the track.

jstebbins commented:
@galad87 that was an unintentional change in behaviour.  I'll fix that.

A little off topic, but related, I've been thinking about the current implementation of passthru and I've got an idea for a follow-up to this patch that simplifies the UI a bit.  Auto-passthru can currently be used to accomplish the same thing as any other individual passthru.  If you want dts passthru, select auto-passthru and check only the dts option in auto-passthru codecs. We have duplicate functionality that usually leads to confusion.  So my thought is, we eliminate all the individual passthru options on the audio defaults pane and only have a "Passthru" that behaves as autopassthru does now. On the regular audio pane, we can also only show "Passthru" (when the source audio can be passed) although in this case it is an implied individual passthru that matches the source audio codec. Does this sound like a worth while thing to do?

jstebbins committed:
Fix fallback behavior for codec specific passthru

Fallback was ignored when using e.g. DTS Passthru

bradleysepos commented:
I think I see where you're going with the passthru stuffs. Would certainly simplify the interface. How would this affect the CLI?

jstebbins commented:
I would leave the CLI alone for now I think.  The underlying support in libhb shouldn't need to change for individual passthrus, so the CLI can continue to use them.  "-E copy:ac3" is a convenient shortcut for "-E copy --audio-copy-mask=ac3". If there is significant simplification we can do in libhb by removing the underlying support for individual passthrus, it's relatively easy to translate the shortcut in test.c.

bradleysepos commented:
Where are we at with this?

Also, I'm concerned about a given scenario. I was under the impression AC3 Passthru would pass through an AC3 source track and convert other types to AC3, so the output type would be guaranteed. Apparently, at some point or in some cases, this was changed to drop the track.

It seems valuable to preserve some way to replicate the original behavior, e.g. for Apple TV which does not support DTS. Usually one wants to convert DTS to AC3 in this case, but still pass through AC3 if possible.

bradleysepos commented:
To clarify, in the Apple TV case, the easy solution is to uncheck the DTS passthru support. But it would be difficult to recreate this scenario if the device did support DTS, and one wanted that option checked. I believe AC3 Passthru did exactly this, and was distinct from Auto Passthru and the fallback mechanism entirely?

galad87 commented:
It still works that way in the mac and linux gui. The win gui doesn't use the same code to apply the audio defaults yet so it might work differently.

bradleysepos commented:
Thanks @galad87.

I think this behavior is worth preserving, and perhaps WinGUI can be updated for parity soon.

sr55 commented:
The WinGui was already changed to not drop the track. It automatically in front-end code switches to the correct encoder since libhb behaviour is to drop the track.

I'll add "None" to the fallback options

jstebbins commented:
If there are no additional comments, this is ready to commit.  TBH, it's been so long, I forgot about this 😁

galad87 commented:
Ship it!
